File: tests/inter-language-test/src/main/cpp/IltConsumerCompressionTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2018 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include "utils/Gmock.h"
#include "utils/Gtest.h"

#include "JoynrTest.h"
#include "joynr/interlanguagetest/TestInterfaceProxy.h"

#include "IltHelper.h"
#include "IltUtil.h"
#include "joynr/JoynrRuntime.h"
#include "joynr/Logger.h"
#include "joynr/MessagingQosEffort.h"

using namespace ::testing;

extern std::string globalIltProgramName;
extern std::string providerDomain;

/** Class to test the compression of joynr messages.
 */
class IltConsumerCompressionTest : public ::testing::Test
{

protected:
    void SetUp()
    {
        // Get the provider domain
        JOYNR_LOG_INFO(logger(), "Creating proxy for provider on domain {}", providerDomain);

        // Get the current program directory
        std::string dir(IltHelper::getAbsolutePathToExecutable(globalIltProgramName));

        // onFatalRuntimeError callback is optional, but it is highly recommended to provide an
        // implementation.
        std::function<void(const joynr::exceptions::JoynrRuntimeException&)> onFatalRuntimeError =
                [&](const joynr::exceptions::JoynrRuntimeException& exception) {
                    JOYNR_LOG_ERROR(
                            logger(),
                            "Unexpected joynr runtime error occured: " + exception.getMessage());
                };

        // Initialize the joynr runtime
        std::string pathToMessagingSettings(dir + "/resources/ilt-consumer.settings");

        runtime = joynr::JoynrRuntime::createRuntime(pathToMessagingSettings, onFatalRuntimeError);

        // Create proxy builder
        proxyBuilder = runtime->createProxyBuilder<joynr::interlanguagetest::TestInterfaceProxy>(
                providerDomain);

        // Messaging Quality of service
        std::int64_t qosMsgTtl = 5000;
        joynr::MessagingQosEffort::Enum qosMsgEffort = joynr::MessagingQosEffort::Enum::NORMAL;
        bool qosMsgEncrypt = false;
        bool qosMsgCompress = true;

        // Find the provider with the highest priority set in ProviderQos
        joynr::DiscoveryQos discoveryQos;
        discoveryQos.setDiscoveryTimeoutMs(60000);
        discoveryQos.setRetryIntervalMs(5000);
        discoveryQos.setCacheMaxAgeMs(std::numeric_limits<std::int64_t>::max());
        discoveryQos.setArbitrationStrategy(
                joynr::DiscoveryQos::ArbitrationStrategy::HIGHEST_PRIORITY);

        // Build a proxy
        testInterfaceProxy =
                proxyBuilder
                        ->setMessagingQos(joynr::MessagingQos(
                                qosMsgTtl, qosMsgEffort, qosMsgEncrypt, qosMsgCompress))
                        ->setDiscoveryQos(discoveryQos)
                        ->build();

        JOYNR_LOG_INFO(logger(), "***********************");
        JOYNR_LOG_INFO(logger(), "Proxy built.");
        JOYNR_LOG_INFO(logger(), "***********************");
    }

    void TearDown()
    {
        testInterfaceProxy.reset();
        proxyBuilder.reset();
        runtime.reset();
    }

    std::shared_ptr<joynr::JoynrRuntime> runtime;
    std::shared_ptr<joynr::interlanguagetest::TestInterfaceProxy> testInterfaceProxy;
    std::shared_ptr<joynr::ProxyBuilder<joynr::interlanguagetest::TestInterfaceProxy>> proxyBuilder;

private:
    ADD_LOGGER(IltConsumerCompressionTest)
};

TEST_F(IltConsumerCompressionTest, callMethodWithoutParameters)
{
    JOYNR_ASSERT_NO_THROW(testInterfaceProxy->methodWithoutParameters());
}

TEST_F(IltConsumerCompressionTest, callMethodWithMultiplePrimitiveParameters)
{
    int32_t arg1 = 2147483647;
    float arg2 = 47.11;
    bool arg3 = false;
    double doubleOut;
    std::string stringOut;
    JOYNR_ASSERT_NO_THROW(testInterfaceProxy->methodWithMultiplePrimitiveParameters(
            doubleOut, stringOut, arg1, arg2, arg3));
    ASSERT_TRUE(IltUtil::cmpDouble(doubleOut, arg2));
    ASSERT_EQ(stringOut, std::to_string(arg1));
}

File: tests/inter-language-test/inter-language-test-base/src/test/java/io/joynr/test/interlanguage/IltConsumerCompressionTest.java
/*
 * #%L
 * %%
 * Copyright (C) 2018 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.test.interlanguage;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import joynr.interlanguagetest.TestInterfaceSync.MethodWithMultiplePrimitiveParametersReturned;

public class IltConsumerCompressionTest extends IltConsumerTest {
    private static final Logger logger = LoggerFactory.getLogger(IltConsumerCompressionTest.class);

    @BeforeClass
    public static void setUp() throws Exception {
        logger.info("setUp: Entering");
        setupConsumerRuntime(true);
        logger.info("setUp: Leaving");
    }

    @AfterClass
    public static void tearDown() throws InterruptedException {
        logger.info("tearDown: Entering");
        generalTearDown();
        logger.info("tearDown: Leaving");
    }

    /*
     * SYNCHRONOUS METHOD CALLS
     */

    // no check possible other than handling exceptions
    @Test
    public void callMethodWithoutParameters() {
        logger.info(name.getMethodName() + "");
        try {
            testInterfaceProxy.methodWithoutParameters();
        } catch (Exception e) {
            fail(name.getMethodName() + " - FAILED: " + e.getMessage());
            return;
        }
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callMethodWithMultiplePrimitiveParameters() {
        logger.info(name.getMethodName() + "");
        try {
            int intArg = 2147483647;
            float floatArg = 47.11f;
            boolean booleanArg = false;
            MethodWithMultiplePrimitiveParametersReturned result;
            result = testInterfaceProxy.methodWithMultiplePrimitiveParameters(intArg, floatArg, booleanArg);
            assertNotNull(result);
            assertTrue(IltUtil.cmpDouble(result.doubleOut, floatArg));
            assertTrue(result.stringOut.equals(Integer.toString(intArg)));
        } catch (Exception e) {
            fail(name.getMethodName() + " - FAILED - caught unexpected exception: " + e.getMessage());
            return;
        }
        logger.info(name.getMethodName() + " - OK");
    }
}
